### ISSUE TITLE
grad is inf/nan when using torch.amp

### ISSUE THREAD

**Issue by taowangcheng** [pytorch/pytorch#111739]
### 🐛 Describe the bug

Below is a very simple for using torch.amp, but the gradients are inf/nan.
```python
import torch
from torch.cuda.amp import GradScaler
from torch import optim
scaler = GradScaler()
a = torch.randn(2, 2, requires_grad=True, device="cuda")
b = torch.randn(2, 2, requires_grad=True, device="cuda")
optimizer = optim.Adam([a, b], lr=0.1)
with torch.autocast(device_type='cuda'):
    c = a @ b
    loss = c.sum()
scaler.scale(loss).backward()
scaler.unscale_(optimizer)
print(a.grad)
```
running results:
```python
tensor([[-inf, nan],
        [-inf, nan]], device='cuda:0')
```

### Versions

```python
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22635-SP0
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Ti Laptop GPU
Nvidia driver version: 531.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2300
DeviceID=CPU0
Family=207
L2CacheSize=14336
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2300
Name=12th Gen Intel(R) Core(TM) i9-12900HX
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.3
[pip3] torch==2.1.0+cu121
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.1.0+cu121
[pip3] torchvision==0.16.0+cu121
[pip3] torchviz==0.0.2
[conda] Could not collect
```

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by jon-chuang** [comment]
From the `GradScaler` docs:

> Therefore, the optimal scale factor is the largest factor that can be used
    without incurring inf or NaN gradient values.
    
This example works with `GradScaler(2.0**15)`

**Comment by taras-janea** [comment]
The reported behavior is [expected](https://github.com/pytorch/pytorch/blob/5e86c0f6289b5b6d673d10e0b1671ff7abc88b97/torch/amp/grad_scaler.py#L53): float16 values can overflow if the scale factor is too large, and the scaler approximates the optimal scale factor over time by checking gradients for `infs`/`NaNs` during `scaler.step()`.

Besides, `torch.cuda.amp.GradScaler(args...)` is now deprecated. The recommended replacement is:
```python
scaler = torch.amp.GradScaler('cuda', init_scale=...)
```
See the [torch.amp docs](https://docs.pytorch.org/docs/stable/amp.html) for more details.

@taowangcheng, please let me know if anything remains to be resolved within the scope of this issue.

**Comment by taras-janea** [comment]
Closing this issue due to inactivity. If you continue to experience this problem or have any related concerns, please don't hesitate to open a new issue.
